### PR TITLE
Don't strip any \0 bytes

### PR DESCRIPTION
**Pull request by ScottPJones**
This fixes the problems I was seeing with \0 bytes being removed from strings.

### CODE REVIEW

**Comment by nalimilan** [comment]
This doesn't pass the tests (though one of the failures is spurious). Are you sure this works somewhere?

**Comment by ScottPJones** [comment]
Sorry, I got called away from finishing pushing the test changes (work problem, upgrade of OS X killed my VPN connection!)
Everything should all work now.
The problem was the inconsistency between `ASCIIString`/`UTF8String` and `UTF16String`/`UTF32String`.
There is an implicit `\0` for the `ByteString` types, when you look at `.data`, but a nul codeunit is *explicit* in the underlying vector `.data` for `UTF16String` and `UTF32String`
(one of the things that causes lots of problems writing low-level generic code that handles all 4 types).

**Comment by ScottPJones** [comment]
I think there is also a bug in the way coverage/coveralls is counting coverage decreased as unsuccessful.  If you remove some lines, but there was no change in which lines were missed,
then that shouldn't count as being failing (the percentage just dropped because total number of lines dropped). cc @kshyatt

**Comment by nalimilan** [comment]
If you remove a lot of lines that were covered, it's logical that the % covered goes down. The metric is quite rudimentary, but at least it's easy to understand. I guess we could find better measures.

File: test/runtests.jl [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
-          "a string € チャネルパートナーの選択")
+          "a string € チャネルパートナーの選択",
+          "a string \0€ チャネルパ\0ー\0トナーの選択 with embedded and trailing nuls\0")
     # Test round-trip to Unicode formats, checking against pure-Julia implementation
-    for T in (UTF8String, UTF16String, UTF32String)
+    for (T, nullen) in ((UTF8String, 0), (UTF16String, 2), (UTF32String, 4))
         enc = iconv.encoding_string(T)
+        # Adjust for explicit \0 only for .data on UTF16String/UTF32String
         a = reinterpret(UInt8, T(s).data)
+        a = a[1:sizeof(a) - nullen]
```
Comment (nalimilan) [inline test/runtests.jl]: Better use `end` here, even if it gives the same result.

**Comment by nalimilan** [comment]
Looks like tests pass. Could you rebase this into two commits, one for the fix, and one for the stylistic changes?

File: test/runtests.jl (hunk 1 above)
Comment (ScottPJones) [inline test/runtests.jl]: Sure!

**Comment by ScottPJones** [comment]
Yes - but if the coverage % drops because of that (esp. given how Julians delight in net code reduction), it really shouldn't be marked as failing the coverage check, should it?
I'm not sure yet where to submit an issue for coverage/coveralls to see if they could fix that.
(or if it's C/C++, try to fix it myself)

**Comment by ScottPJones** [comment]
Hopefully the commits are how you wanted now.
The problems with using `.data` is why I think it would be nice to have a `codeunit` method, that would return a `Vector` of `UInt8`,`UInt16`,`UInt32` for all 4 types, without the inconsistent `\0`. Then `reinterpret(UInt8, codeunit(str))` would always give what you wanted for these encode/decode functions.

**Comment by nalimilan** [comment]
> Hopefully the commits are how you wanted now.

Ah, please squash the two first commit: it's not a good idea to have a commit that doesn't pass the tests in the build history (bisecting...).

> The problems with using .data is why I think it would be nice to have a codeunit method, that would return a Vector of UInt8,UInt16,UInt32 for all 4 types, without the inconsistent \0. Then reinterpret(UInt8, codeunit(str)) would always give what you wanted for these encode/decode functions.

Yeah, we'll have to consider this kind of issue once Stefan's new `String` type will have been merged.

**Comment by ScottPJones** [comment]
Done!
>Yeah, we'll have to consider this kind of issue once Stefan's new String type will have been merged.

If that gets merged, I may end up having to maintain a separate fork or something, because of the performance implications and code breakage for people using UTF16 and UTF32 (and anyone going back and forth between `Char` and `UInt32` a lot).  I hope that enough string performance benchmarks are added (hopefully with the new performance benchmarker) so that the part of changing `Char` to be some packed form of `UTF8` is dropped form the PR. (I've been tracking the highlander branch for many months now)

**Comment by ScottPJones** [comment]
BTW, what do you feel about changing the errors to use something like I did for `UnicodeError`?
It can help possibly help performance, because it might eliminate the need to build a GC frame in some cases (if it is just passing a code and some integer values to the exception constructor, instead of constructing a string for `error` as currently)

**Comment by nalimilan** [comment]
> If that gets merged, I may end up having to maintain a separate fork or something, because of the performance implications and code breakage for people using UTF16 and UTF32 (and anyone going back and forth between Char and UInt32 a lot). I hope that enough string performance benchmarks are added (hopefully with the new performance benchmarker) so that the part of changing Char to be some packed form of UTF8 is dropped form the PR. (I've been tracking the highlander branch for many months now)

I guess you'll simply have to create your own concrete subtype of `AbstractChar`, that's all.



> BTW, what do you feel about changing the errors to use something like I did for UnicodeError?
> It can help possibly help performance, because it might eliminate the need to build a GC frame in some cases (if it is just passing a code and some integer values to the exception constructor, instead of constructing a string for error as currently)

Sure, I've always wanted to improve the current system.

**Comment by ScottPJones** [comment]
Thanks!

**Pull request merged**